wire/internal/wire/wire.go

```diff
@@ -661,22 +661,35 @@ func zeroValue(t types.Type, qf types.Qualifier) string {
 	}
 }
 
-// typeVariableName invents a variable name derived from the type name
-// or returns the empty string if one could not be found. There are no
-// guarantees about whether the name is exported or unexported: call
-// export() or unexport() to convert.
-func typeVariableName(t types.Type) string {
+// typeVariableName invents variable names derived from the type name,
+// or returns a slice with defaultName if one could not be found.
+// The first returned name is preferred; others may be used for
+// disambiguation.
+// There are no guarantees about whether the names are exported or unexported:
+// call export() or unexport() to convert.
+func typeVariableName(t types.Type, defaultName string) []string {
```

Comment:
While I like that this is composable, I think it might push too much complexity on the caller (and thus spreading the places that need to be tested). I think it's time for `typeVariableName` to return the final name directly. Perhaps the signature could be:

```go
func typeVariableName(t types.Type, defaultName string, transform func(string) string, collides func(string) bool) string
```

Reply:
Done. Also added a test for typeVariableName.